rollbar/client.go

```diff
@@ -0,0 +1,149 @@
+package rollbar
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+const apiURL = "https://api.rollbar.com/api/1"
+
+// Client represents the rollbar client.
+type Client struct {
+	APIKey      string
+	APIScheme   string
+	APIHost     string
+	APIBasePath string
+}
+
+// Option adds the base url and other parameters to the client.
+type Option func(*Client) error
+
+// NewClient is a constructor.
+func NewClient(apiKey string, opts ...Option) (*Client, error) {
+	url, err := url.Parse(apiURL)
+	if err != nil {
+		return nil, err
+	}
+
+	client := Client{
+		APIKey:      apiKey,
+		APIScheme:   url.Scheme,
+		APIHost:     url.Host,
+		APIBasePath: url.Path,
+	}
+
+	if err = client.parseOptions(opts...); err != nil {
+		return nil, err
+	}
+
+	return &client, nil
+}
+
+func (c *Client) parseOptions(opts ...Option) error {
+	// Range over each options function and apply it to our API type to
+	// configure it. Options functions are applied in order, with any
+	// conflicting options overriding earlier calls.
+	for _, option := range opts {
+		err := option(c)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// BaseURL returns a function for configuring the url inside a Client.
+func BaseURL(baseURL string) Option {
+	return func(c *Client) error {
+		var apiURL = &url.URL{}
+		apiURL, err := url.Parse(baseURL)
+		if err != nil {
+			return err
+		}
+
+		c.APIBasePath = apiURL.Path
+		if apiURL.Path == "/" {
+			c.APIBasePath = ""
+		}
+
+		c.APIHost = apiURL.Host
+		c.APIScheme = apiURL.Scheme
+
+		return nil
+	}
+}
+
+func (c *Client) get(pathComponents ...string) ([]byte, error) {
+	return c.getWithQueryParams(map[string]string{}, pathComponents...)
+}
+
+func (c *Client) getWithQueryParams(queryParams map[string]string, pathComponents ...string) ([]byte, error) {
+	url := c.url(true, queryParams, pathComponents...)
+	return c.makeRequest("GET", url, nil)
+}
+
+func (c *Client) post(data []byte, pathComponents ...string) ([]byte, error) {
+	url := c.url(false, map[string]string{}, pathComponents...)
+	body := bytes.NewBuffer(data)
+	return c.makeRequest("POST", url, body)
+}
+
+func (c *Client) delete(pathComponents ...string) error {
+	url := c.url(true, map[string]string{}, pathComponents...)
+	_, err := c.makeRequest("DELETE", url, nil)
+	return err
+}
+
+func (c *Client) url(withAccessToken bool, queryMap map[string]string, pathComponents ...string) string {
+	query := url.Values{}
+	for key, value := range queryMap {
+		query.Add(key, value)
+	}
+
+	if withAccessToken {
+		query.Add("access_token", c.APIKey)
+	}
+
+	components := append([]string{c.APIBasePath}, pathComponents...)
```

Comment:
I see that `APIBasePath` has a type `string`, so not sure, why do you type cast it explicitly again?

Reply:
@graffzon This is not type casting to string, but the constructor of `[]string` with assignment.